unison-cli/src/Unison/Cli/Monad.hs

```diff
@@ -0,0 +1,349 @@
+    ucmVersion :: UCMVersion
+  }
+  deriving stock (Generic)
+
+-- | The command-line app monad mutable state.
+--
+-- Get the state with 'getLoopState', put it with 'putLoopState', or modify it with 'modifyLoopState'.
+--
+-- There's an additional pseudo @"currentPath"@ field lens, for convenience.
+data LoopState = LoopState
+  { root :: Branch IO,
+    lastSavedRoot :: Branch IO,
+    -- the current position in the namespace
+    currentPathStack :: List.NonEmpty Path.Absolute,
+    -- TBD
+    -- , _activeEdits :: Set Branch.EditGuid
+
+    -- The file name last modified, and whether to skip the next file
+    -- change event for that path (we skip file changes if the file has
+    -- just been modified programmatically)
+    latestFile :: Maybe (FilePath, Bool),
+    latestTypecheckedFile :: Maybe (UF.TypecheckedUnisonFile Symbol Ann),
+    -- The previous user input. Used to request confirmation of
+    -- questionable user commands.
+    lastInput :: Maybe Input,
+    -- A 1-indexed list of strings that can be referenced by index at the
+    -- CLI prompt.  e.g. Given ["Foo.bat", "Foo.cat"],
+    -- `rename 2 Foo.foo` will rename `Foo.cat` to `Foo.foo`.
+    numberedArgs :: NumberedArgs
+  }
+  deriving stock (Generic)
+
+instance
+  {-# OVERLAPS #-}
+  Functor f =>
+  IsLabel "currentPath" ((Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState))
+  where
+  fromLabel :: (Path.Absolute -> f Path.Absolute) -> (LoopState -> f LoopState)
+  fromLabel =
+    lens
+      (\LoopState {currentPathStack} -> List.NonEmpty.head currentPathStack)
+      ( \loopState@LoopState {currentPathStack = _ List.NonEmpty.:| paths} path ->
+          loopState {currentPathStack = path List.NonEmpty.:| paths}
+      )
+
+-- | Create an initial loop state given a root branch and the current path.
+loopState0 :: Branch IO -> Path.Absolute -> LoopState
+loopState0 b p =
+  LoopState
+    { root = b,
+      lastSavedRoot = b,
+      currentPathStack = pure p,
+      latestFile = Nothing,
+      latestTypecheckedFile = Nothing,
+      lastInput = Nothing,
+      numberedArgs = []
+    }
+
+-- | Run a @Cli@ action down to @IO@.
+runCli :: Env -> Cli a a -> IO (ReturnType a)
+runCli env (Cli action) =
+  action env (\x -> pure (Success x))
+
+-- | Get the loop state.
+getLoopState :: Cli r LoopState
+getLoopState = do
+  Env {loopStateRef} <- ask
+  liftIO (readIORef loopStateRef)
+
+-- | Put the loop state.
+putLoopState :: LoopState -> Cli r ()
+putLoopState newSt = do
+  Env {loopStateRef} <- ask
+  liftIO (writeIORef loopStateRef newSt)
+
+-- | Modify the loop state (not thread-safe).
+modifyLoopState :: (LoopState -> LoopState) -> Cli r ()
+modifyLoopState f = do
+  Env {loopStateRef} <- ask
+  liftIO (modifyIORef' loopStateRef f)
+
+-- | The result of calling 'loadSource'.
+data LoadSourceResult
+  = InvalidSourceNameError
+  | LoadError
+  | LoadSuccess Text
+
+-- | Lift an action of type @IO (Either e a)@, given a continuation for @e@.
```

Comment:
Does this combinator provide anything special over simply running the action and pattern-matching on the result?

e.g.

```
action >>= \case
  Left e -> doSomething
  Right a -> doSomethingElse
```

?

I ask because I wouldn't expect to need to `unCli` and such to do this sort of thing; seems more complex that necessary so I'm guessing I'm missing something 🤔